Tutorial app

notlfish opened the pull request:
### What?
Pull request to implement the app developed through [this react tutorial](https://ibaslogic.com/react-tutorial-for-beginners/).

### Before merging
React components implementing the following functions are to be completed before merging into `develop`:
- [x] Display app header
- [x] Handle input form
- [x] Display todo list
- [x] Check completed tasks
- [x] Delete todo
- [x] Update todo description
- [x] Fetch tasks from JSONplaceholder API
- [x] Persist list using `localStorage`
- [x] Add routing for Home, About, and Error pages

notlfish committed:
Add TodoContainer.js file

notlfish committed:
Add TodosList and TodoItem components

notlfish committed:
Add Header component

notlfish committed:
Add checkbox to TodoItem components

notlfish committed:
Add checkbox handle function to TodoContainer component

notlfish committed:
Complete checkbox button and add delete button

notlfish committed:
Add task input form and styling

notlfish committed:
Add task editing functionality

notlfish committed:
Add functionality to abort editing tasks

notlfish committed:
Persist todo list into localStorage

notlfish committed:
Fetch initial Todos from JSONPlaceholder API

notlfish committed:
Add About and Error pages, and routing

notlfish committed:
Add about info with dynamic routing

notlfish committed:
Add button icons

notlfish committed:
Fix linting errors

enionsouza commented on src/components/InputTodo.js:
```diff
@@ -0,0 +1,59 @@
+    };
+  }
+
+  onChange = (e) => {
+    this.setState({
+      [e.target.name]: e.target.value,
+    });
```
- [x] If the only field of `inputText` is `title`, then it may be better to just do something like 

```javascript
setInputText({
  title: e.target.value,
});
```
Reason: it's more explicit and easier for someone reading the code to know what's happening.

notlfish replied:
Agreed, I'll be changing this in the next commit

enionsouza commented on src/components/TodoItem.js:
```diff
@@ -0,0 +1,106 @@
+        editing: false,
+      });
+    }
+    if (event.key === 'Escape') {
+      this.setState({
+        editing: false,
+      });
+    }
+  }
```
Please, remove the changes to the "buffer title" after the user has chosen to Escape the Edition Mode.

notlfish replied:
Oh, totally. I also like that behavior better, just didn't think about it the time around 😅

enionsouza commented on src/components/TodosList.js:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+TodosList.propTypes = {
+  todos: PropTypes.arrayOf(
+    PropTypes.shape({
+      id: PropTypes.string.isRequired,
+      title: PropTypes.string.isRequired,
+      completed: PropTypes.bool.isRequired,
+    }),
+  ).isRequired,
+  handleChangeProps: PropTypes.func.isRequired,
+  handleDelete: PropTypes.func.isRequired,
+  setUpdate: PropTypes.func.isRequired,
+};
```
You did an excellent job on checking your Prop Types! :+1: :100:

notlfish replied:
:grin: maybe it would be nice to have these PropTypes defined in a single place of the project? Particularly the ones representing a todo.

enionsouza reviewed (changes requested):
### Changes Required...
Your project is awesome at this point, and it is fantastic to see how you changed the styles for the Navbar component in such an interesting way! :+1: :100: :1st_place_medal:
![](https://media2.giphy.com/media/3oEjI5VtIhHvK37WYo/giphy.gif)

Please consider the two comments that I have made about your code. The third comment is just a remark on your mastering of PropTypes!

notlfish committed:
Implemented changes requested on review.

- Simplified state update of InputTodo#onChange to explicitly change the title.
- Changed behavior when escaping a TodoItem editing: now discards unsaved changes.

enionsouza reviewed (approved):
### Congratulations! Your project has been approved!
![](https://media2.giphy.com/media/NwzwmG2UFQlwmfwUgM/giphy.gif)
